### PR TITLE
Add onlyForGroups parameter to @BeforeMethod

### LINKED ISSUE #549: Groups in @BeforeMethod and @AfterMethod don't work as expected
Either I'm misinterpreting or misusing TestNG-6.8.8, but it seems like the groups attribute in the @BeforeMethod and @AfterMethod aren't working as expected.

Consider following class:

```java
public class MyClass {
    public void method1() {
        System.out.println("Method 1 invocation");
    }
    public void method2() {
        System.out.println("Method 2 invocation");
    }
}
```

and following test

```java
public class MyClassTest {
    MyClass o = new MyClass();
    @BeforeMethod(groups = "method1") 
    public void before1() {
        System.out.println("Before1");
    }
    @BeforeMethod(groups = "method2") 
    public void before2() {
        System.out.println("Before2");
    }

    @AfterMethod(groups = "method1") 
    public void after1() {
        System.out.println("After1");
    }
    @AfterMethod(groups = "method2") 
    public void after2() {
        System.out.println("After2");
    }

    @Test(groups = "method1")
    public void testMethod1() throws Exception {
        o.method1();
    }
    @Test(groups = "method2")
    public void testMethod2() throws Exception {
        o.method2();
    }
}
```

I would expect the output to be
```
Before1
Method 1 invocation
After1
Before2
Method 2 invocation
After2
```

However, the output is
```
Before1
Before2
Method 1 invocation
After1
After2
Before1
Before2
Method 2 invocation
After1
After2
```

### PR DESCRIPTION
**Pull request by stachenov**
It allows to run @BeforeMethod configuration methods only if the test
method being invoked belongs to one of the listed groups.

I hope it fixes #549 and #780.

It certainly does need further refactoring and javadoc updating, but I'm postponing it until the general idea is approved.

Todo:
- [x] Put the new tests into `testng.xml`, but where exactly?
- [x] Maybe move those tests into another package?
- [ ] Update CHANGES.txt.
- [ ] Do the same thing for `@AfterMethod` once `@BeforeMethod` is done with.

### CODE REVIEW

File: src/test/java/test/configuration/BeforeMethodWithGroupFiltersTest.java
Comment (krmahadevan) [inline src/test/java/test/configuration/BeforeMethodWithGroupFiltersTest.java]: Please add this to `src/test/resources/testng.xml`

File: src/main/java/org/testng/annotations/BeforeMethod.java
Comment (krmahadevan) [inline src/main/java/org/testng/annotations/BeforeMethod.java]: Can you please add some java docs to this attribute, that explains what this attribute can be used for ?

File: src/main/java/org/testng/internal/ConfigurationMethod.java
Comment (krmahadevan) [inline src/main/java/org/testng/internal/ConfigurationMethod.java]: Can we please have this variable initialized to an empty array so that we can avoid a null check to be done ?

**Review by krmahadevan — CHANGES_REQUESTED** [review]
Also please help update the `CHANGES.txt` referring to the set of bugs that is being fixed by this PR. If this is a new functionality please update the text file accordingly.

File: src/test/java/test/configuration/BeforeMethodWithGroupFiltersTest.java
Comment (krmahadevan) [reply src/test/java/test/configuration/BeforeMethodWithGroupFiltersTest.java]: Maybe into `<test name="Configuration">` ?

File: src/test/java/test/configuration/BeforeMethodWithGroupFiltersSampleTest.java
Comment (juherr) [inline src/test/java/test/configuration/BeforeMethodWithGroupFiltersSampleTest.java]: `g2m2`?

File: src/test/java/test/configuration/BeforeMethodWithGroupFiltersSampleTest.java
Comment (juherr) [inline src/test/java/test/configuration/BeforeMethodWithGroupFiltersSampleTest.java]: I prefer the InvokedMethodListener instead of static fields.

File: src/test/java/test/configuration/BeforeMethodWithGroupFiltersTest.java
Comment (juherr) [inline src/test/java/test/configuration/BeforeMethodWithGroupFiltersTest.java]: the assertion is not strong enough: the order should not change so you can provide the exact list of invoked methods.

**Review by juherr — CHANGES_REQUESTED** [review]

**Comment by stachenov** [comment]
OK, I think I've made all the requested changes. I have one doubt, though: in the process, I had to move `getGroupFilters` from `IConfigurationAnnotation` to `IBeforeMethod`, which makes sense. But then in `AnnotationHelper.finishInitialize` I had to remove the appropriate call because the `bs` parameter has the `IConfigurationAnnotation` type, which means I can't call `getGroupFilters` on it any more. The tests still pass, and I can't seem to understand what that `finishInitialize` method is for. Looks like a poor man's copy constructor that is totally out of place to me.

Other than that, I think I'm done with `@BeforeMethod`. Should I move on to `@AfterMethod`? I plan to update CHANGES.txt as soon as both are done because it seems like one logical change to me.

File: src/main/java/org/testng/collections/Sets.java
Comment (juherr) [inline src/main/java/org/testng/collections/Sets.java]: Just a detail: param could be `V...`

File: src/main/java/org/testng/internal/ConfigurationMethod.java
Comment (juherr) [inline src/main/java/org/testng/internal/ConfigurationMethod.java]: Again a detail and it's just a personal taste but I prefer to see the business code at the root level of the method
```java
   IBeforeMethod before = m_annotationFinder.findAnnotation(getConstructorOrMethod(), IBeforeMethod.class);
   if (before == null) {
     return new String[0];
   }
   return before.getGroupFilters();
```

File: src/test/java/test/configuration/BeforeMethodWithGroupFiltersTest.java
Comment (juherr) [inline src/test/java/test/configuration/BeforeMethodWithGroupFiltersTest.java]: If `run` has varargs, no need to create the array.
And if `run` doesn't have it, you can modify it.

File: src/test/java/test/configuration/BeforeMethodWithGroupFiltersTest.java
Comment (juherr) [inline src/test/java/test/configuration/BeforeMethodWithGroupFiltersTest.java]: To check, but I think `isEqualTo` is not strict enough and/or won't provide the best feedback in case of wrong assertion. `containsExactly` looks better.

**Review by juherr — APPROVED** [review]
Some minor remarks

**Comment by juherr** [comment]
@stachenov The process is pretty simple: if all the tests pass and the api is backward compatible, it's ok.

**Comment by juherr** [comment]
@stachenov About `@AfterMethod`, please wait for @cbeust feedback on the PR first because it changes a bit the current behavior (even if I think it is a good change).

**Comment by stachenov** [comment]
Very reasonable remarks, updated the code accordingly. One thing I don't understand: how does it change the current behavior? As long as nobody uses the new attribute, everything should work exactly the same way as before.

**Comment by krmahadevan** [comment]
@stachenov - It would be great if you could please help squash your commits and push them as one single commit.

**Comment by juherr** [comment]
@stachenov We can imagine some person liked the previous behavior.

**Comment by stachenov** [comment]
@krmahadevan I was going to squash them, but when exactly should I do it? Now, or should I get done with `@AfterMethod` first and then squash everything? Also, I'm a git newbie, so just to be safe—if I just `rebase -i` my branch to squash everything and then `push --force-with-lease` it, is it OK? Won't I break anything, will this PR still be valid?

@juherr I still don't get it. If someone liked the previous behavior, then they are free to ignore the new attribute and never use it, nothing will change for them at all. I can hardly imagine anyone being annoyed by a new feature that is hardly noticeable unless you look for it.

**Comment by cbeust** [comment]
> in the process, I had to move getGroupFilters from IConfigurationAnnotation to IBeforeMethod

Mmmh, if this interface is part of the public API, this is a breaking change.

**Comment by stachenov** [comment]
> Mmmh, if this interface is part of the public API, this is a breaking change.

Oops, I hadn't thought of that. But it turns out that exactly this action turned a breaking change into a safe one. `IConfigurationAnnotation`, where I initially put the new method, is indeed a part of the public API. `IBeforeMethod`, where I ended up moving it to, on the other hand, is a part of the `org.testng.internal` package, so I guess it's not a public API.

As far as I can see now, I only touched the public API in two places: added a new attribute to the `@BeforeMethod` annotation and a new static utility method to the `Sets` utility class. Both seem safe to me.

**Comment by juherr** [comment]
@stachenov 

> If someone liked the previous behavior, then they are free to ignore the new attribute and never use it

Right, I forgot the new attribute!

@cbeust I think this API modification is acceptable because it's just an addition and it will be part of the next major release.

**Comment by krmahadevan** [comment]
@stachenov - Looks like @cbeust already merged your code. For your next delivery, after your initial push, you can consider amending your commit to address review comments. Since you mentioned you are a git newbie, perhaps you can look at this [post](https://blog.carbonfive.com/2017/08/28/always-squash-and-rebase-your-git-commits/)

**Comment by stachenov** [comment]
@krmahadevan I thought about amending, but commit history is certainly useful during discussion. This way I can look back at what I did and why at any point, which is especially useful if it drags on long enough. Is `rebase -i` before final merge a wrong way to squash everything? I know *how* to squash commits, I just have to learn when and why do/don't it.

I think it was merged too soon. It's a minor thing that I haven't updated CHANGES.txt yet, but what about the same mechanism for `@AfterMethod`? Should I open another PR for that?

**Comment by krmahadevan** [comment]
@stachenov - Yes, you can create a new PR for `@AfterMethod` and in that PR you should definitely update `CHANGES.txt`.

Also while you send out the PR for `@AfterMethod`, request you to please help update our user facing documentation also with this new feature.

For doing that, you should raise a PR in this repo : https://github.com/testng-team/testng-team.github.io 

That way http://testng.org/doc/documentation-main.html gets updated with this details

**Comment by nataliyatest** [comment]
@stachenov seems this annotation is still missing for BeforeMethod/AfterMethod in v6.14.3
image

**Comment by juherr** [comment]
@nataliyatest It will be available for the next release and you can already test it with the snapshot version

**Pull request merged**
